Sources/AsyncHTTPClient/FileDownloadDelegate.swift

```diff
@@ -0,0 +1,121 @@
+
+    private var fileHandleFuture: EventLoopFuture<NIOFileHandle>?
+    private var writeFuture: EventLoopFuture<Void>?
+
+    /// Initializes a new file download delegate.
+    /// - parameters:
+    ///     - path: Path to a file you'd like to write the download to.
+    ///     - pool: A thread pool to use for asynchronous file I/O.
+    ///     - reportHeaders: A closure called when the response headers are available.
+    ///     - reportProgress: A closure called when a body chunk has been downloaded, with
+    ///       the total byte count and download byte count passed to it as arguments. The callbacks
+    ///       will be invoked in the same threading context that the delegate itself is invoked,
+    ///       as controlled by `EventLoopPreference`.
+    public init(
+        path: String,
+        pool: NIOThreadPool = NIOThreadPool(numberOfThreads: 1),
+        reportHeaders: ((HTTPHeaders) -> Void)? = nil,
+        reportProgress: ((Progress) -> Void)? = nil
+    ) throws {
+        pool.start()
+        self.io = NonBlockingFileIO(threadPool: pool)
+        self.filePath = path
+
+        self.reportHeaders = reportHeaders
+        self.reportProgress = reportProgress
+    }
+
+    public func didReceiveHead(
+        task: HTTPClient.Task<Response>,
+        _ head: HTTPResponseHead
+    ) -> EventLoopFuture<Void> {
+        self.reportHeaders?(head.headers)
+
+        if let totalBytesString = head.headers.first(name: "Content-Length"),
+            let totalBytes = Int(totalBytesString) {
+            self.progress.totalBytes = totalBytes
+        }
+
+        return task.eventLoop.makeSucceededFuture(())
+    }
+
+    public func didReceiveBodyPart(
+        task: HTTPClient.Task<Response>,
+        _ buffer: ByteBuffer
+    ) -> EventLoopFuture<Void> {
+        self.progress.receivedBytes += buffer.readableBytes
+        self.reportProgress?(self.progress)
+
+        let writeFuture: EventLoopFuture<Void>
+        if let fileHandleFuture = self.fileHandleFuture {
+            writeFuture = fileHandleFuture.flatMap {
+                self.io.write(fileHandle: $0, buffer: buffer, eventLoop: task.eventLoop)
+            }
+        } else {
+            let fileHandleFuture = self.io.openFile(
+                path: self.filePath,
+                mode: .write,
+                flags: .allowFileCreation(),
+                eventLoop: task.eventLoop
+            )
+            self.fileHandleFuture = fileHandleFuture
+            writeFuture = fileHandleFuture.flatMap {
+                self.io.write(fileHandle: $0, buffer: buffer, eventLoop: task.eventLoop)
+            }
+        }
+
+        self.writeFuture = writeFuture
+        return writeFuture
+    }
+
+    private func close(fileHandle: NIOFileHandle) {
+        try! fileHandle.close()
+        self.fileHandleFuture = nil
+    }
+
+    public func didFinishRequest(task: HTTPClient.Task<Response>) throws -> Response {
```

Comment:
Can we add a test on the error path? I think we're missing cleanup in the `didReceiveError` case.